File: src/google/protobuf/incremental_parser.cc
// Protocol Buffers - Google's data interchange format
// Copyright 2008 Google Inc.  All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file or at
// https://developers.google.com/open-source/licenses/bsd

#include "google/protobuf/incremental_parser.h"

#include <algorithm>
#include <cstddef>
#include <cstdint>
#include <cstring>
#include <utility>

#include "absl/strings/string_view.h"
#include "absl/types/span.h"

namespace google {
namespace protobuf {

IncrementalParser::IncrementalParser()
    : IncrementalParser(Options{}) {}

IncrementalParser::IncrementalParser(const Options& options)
    : options_(options),
      state_(State::kNeedMoreData),
      buffer_pos_(0),
      bytes_consumed_(0),
      message_size_(0),
      have_message_size_(false) {
  buffer_.reserve(options_.initial_buffer_capacity);
}

IncrementalParser::~IncrementalParser() = default;

IncrementalParser::IncrementalParser(IncrementalParser&&) noexcept = default;
IncrementalParser& IncrementalParser::operator=(IncrementalParser&&) noexcept = default;

IncrementalParser::State IncrementalParser::Feed(absl::Span<const char> bytes) {
  if (state_ == State::kError || state_ == State::kDone) {
    return state_;
  }

  // Append bytes to buffer
  if (!bytes.empty()) {
    buffer_.insert(buffer_.end(), bytes.begin(), bytes.end());
  }

  // Try to parse a message
  TryParse();

  return state_;
}

IncrementalParser::State IncrementalParser::Feed(absl::string_view bytes) {
  return Feed(absl::Span<const char>(bytes.data(), bytes.size()));
}

bool IncrementalParser::TryParse() {
  if (options_.length_delimited) {
    // Parse length-delimited messages
    if (!have_message_size_) {
      // Try to read the message size
      uint32_t size;
      if (!ReadVarint(&size)) {
        state_ = State::kNeedMoreData;
        return false;
      }

      if (size > options_.max_message_size) {
        state_ = State::kError;
        error_message_ = "Message size exceeds maximum allowed size";
        return false;
      }

      message_size_ = size;
      have_message_size_ = true;
    }

    // Check if we have enough bytes for the message
    size_t available = buffer_.size() - buffer_pos_;
    if (available < message_size_) {
      state_ = State::kNeedMoreData;
      return false;
    }

    // Extract the message bytes
    message_bytes_.assign(buffer_.data() + buffer_pos_, message_size_);
    buffer_pos_ += message_size_;
    bytes_consumed_ += message_size_;

    // Compact buffer if we've consumed a lot
    if (buffer_pos_ > buffer_.size() / 2 && buffer_pos_ > 4096) {
      buffer_.erase(buffer_.begin(), buffer_.begin() + buffer_pos_);
      buffer_pos_ = 0;
    }

    state_ = State::kMessageReady;
    return true;
  } else {
    // Non-length-delimited: parse entire buffer as a single message
    if (buffer_.empty()) {
      state_ = State::kNeedMoreData;
      return false;
    }

    message_bytes_.assign(buffer_.begin() + buffer_pos_, buffer_.end());
    bytes_consumed_ += message_bytes_.size();
    buffer_pos_ = buffer_.size();
    state_ = State::kMessageReady;
    return true;
  }
}

bool IncrementalParser::ReadVarint(uint32_t* value) {
  size_t pos = buffer_pos_;
  uint32_t result = 0;
  int shift = 0;

  while (pos < buffer_.size()) {
    uint8_t byte = static_cast<uint8_t>(buffer_[pos]);
    result |= static_cast<uint32_t>(byte & 0x7F) << shift;
    pos++;

    if ((byte & 0x80) == 0) {
      // Varint complete
      bytes_consumed_ += (pos - buffer_pos_);
      buffer_pos_ = pos;
      *value = result;
      return true;
    }

    shift += 7;
    if (shift >= 35) {
      // Varint too long
      state_ = State::kError;
      error_message_ = "Varint is too long";
      return false;
    }
  }

  // Need more data
  return false;
}

void IncrementalParser::Reset() {
  state_ = State::kNeedMoreData;
  buffer_.clear();
  buffer_pos_ = 0;
  bytes_consumed_ = 0;
  message_size_ = 0;
  have_message_size_ = false;
  error_message_.clear();
  message_bytes_.clear();
}

void IncrementalParser::Finish() {
  if (state_ == State::kNeedMoreData && buffer_pos_ >= buffer_.size()) {
    state_ = State::kDone;
  }
}

}  // namespace protobuf
}  // namespace google

File: java/core/src/main/java/com/google/protobuf/IncrementalParser.java
// Protocol Buffers - Google's data interchange format
// Copyright 2008 Google Inc.  All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file or at
// https://developers.google.com/open-source/licenses/bsd

package com.google.protobuf;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;

/**
 * IncrementalParser provides a state machine for parsing protocol buffer messages
 * incrementally as bytes become available. This is useful for streaming scenarios
 * where data arrives in chunks.
 *
 * <p>Example usage:
 * <pre>{@code
 * IncrementalParser<Person> parser = new IncrementalParser<>(Person.parser());
 * while (true) {
 *     byte[] bytes = network.read();
 *     IncrementalParser.State state = parser.feed(bytes);
 *     if (state == IncrementalParser.State.MESSAGE_READY) {
 *         Person person = parser.takeMessage();
 *         processMessage(person);
 *     }
 * }
 * }</pre>
 *
 * @param <T> the message type being parsed
 */
public class IncrementalParser<T extends Message> {

  /** Parser states */
  public enum State {
    /** Parser needs more bytes to complete message */
    NEED_MORE_DATA,
    /** A complete message is ready to be taken */
    MESSAGE_READY,
    /** An error occurred during parsing */
    ERROR,
    /** Parsing is complete (no more messages expected) */
    DONE
  }

  private final Parser<T> messageParser;
  private final ByteArrayOutputStream buffer;
  private final long maxMessageSize;

  private State state;
  private int messageSize;
  private boolean haveMessageSize;
  private long bytesConsumed;
  private String errorMessage;
  private T parsedMessage;

  // Varint parsing state
  private int varintValue;
  private int varintShift;

  /** Default maximum message size (64MB) */
  public static final long DEFAULT_MAX_MESSAGE_SIZE = 64 * 1024 * 1024;

  /**
   * Creates an IncrementalParser with default settings.
   *
   * @param parser the parser for the message type
   */
  public IncrementalParser(Parser<T> parser) {
    this(parser, DEFAULT_MAX_MESSAGE_SIZE);
  }

  /**
   * Creates an IncrementalParser with a custom maximum message size.
   *
   * @param parser the parser for the message type
   * @param maxMessageSize maximum allowed message size in bytes
   */
  public IncrementalParser(Parser<T> parser, long maxMessageSize) {
    this.messageParser = parser;
    this.maxMessageSize = maxMessageSize;
    this.buffer = new ByteArrayOutputStream();
    this.state = State.NEED_MORE_DATA;
    this.messageSize = 0;
    this.haveMessageSize = false;
    this.bytesConsumed = 0;
    this.varintValue = 0;
    this.varintShift = 0;
  }

  /**
   * Feed bytes to the parser incrementally.
   *
   * @param bytes the bytes to feed
   * @return the current state after processing the bytes
   */
  public State feed(byte[] bytes) {
    return feed(bytes, 0, bytes.length);
  }

  /**
   * Feed bytes to the parser incrementally.
   *
   * @param bytes the byte array containing data
   * @param offset the offset in the array
   * @param length the number of bytes to process
   * @return the current state after processing the bytes
   */
  public State feed(byte[] bytes, int offset, int length) {
    if (state == State.ERROR || state == State.DONE) {
      return state;
    }

    int pos = offset;
    int end = offset + length;

    while (pos < end) {
      if (!haveMessageSize) {
        // Parse the varint size prefix
        while (pos < end) {
          byte b = bytes[pos++];
          varintValue |= (b & 0x7F) << varintShift;

          if ((b & 0x80) == 0) {
            // Varint complete
            messageSize = varintValue;
            varintValue = 0;
            varintShift = 0;

            if (messageSize > maxMessageSize) {
              state = State.ERROR;
              errorMessage = "Message size exceeds maximum: " + messageSize;
              return state;
            }

            haveMessageSize = true;
            break;
          }

          varintShift += 7;
          if (varintShift >= 35) {
            state = State.ERROR;
            errorMessage = "Varint too long";
            return state;
          }
        }
      }

      if (haveMessageSize) {
        // Read message bytes
        int remaining = messageSize - buffer.size();
        int available = end - pos;
        int toRead = Math.min(remaining, available);

        if (toRead > 0) {
          buffer.write(bytes, pos, toRead);
          pos += toRead;
          bytesConsumed += toRead;
        }

        if (buffer.size() >= messageSize) {
          // Message complete, parse it
          try {
            parsedMessage = messageParser.parseFrom(buffer.toByteArray());
            state = State.MESSAGE_READY;
            return state;
          } catch (InvalidProtocolBufferException e) {
            state = State.ERROR;
            errorMessage = "Failed to parse message: " + e.getMessage();
            return state;
          }
        }
      }
    }

    state = State.NEED_MORE_DATA;
    return state;
  }

  /**
   * Feed bytes from a ByteBuffer to the parser.
   *
   * @param buffer the ByteBuffer to read from
   * @return the current state after processing the bytes
   */
  public State feed(ByteBuffer buffer) {
    if (buffer.hasArray()) {
      return feed(buffer.array(), buffer.arrayOffset() + buffer.position(), buffer.remaining());
    } else {
      byte[] bytes = new byte[buffer.remaining()];
      buffer.get(bytes);
      return feed(bytes);
    }
  }

  /**
   * Get the current parser state.
   *
   * @return the current state
   */
  public State getState() {
    return state;
  }

  /**
   * Check if a complete message is ready to be taken.
   *
   * @return true if a message is ready
   */
  public boolean hasMessage() {
    return state == State.MESSAGE_READY;
  }

  /**
   * Take the parsed message. The parser will reset for the next message.
   *
   * @return the parsed message, or null if no message is ready
   */
  public T takeMessage() {
    if (state != State.MESSAGE_READY) {
      return null;
    }

    T message = parsedMessage;
    parsedMessage = null;
    buffer.reset();
    haveMessageSize = false;
    messageSize = 0;
    state = State.NEED_MORE_DATA;

    return message;
  }

  /**
   * Get the number of bytes consumed so far.
   *
   * @return bytes consumed
   */
  public long getBytesConsumed() {
    return bytesConsumed;
  }

  /**
   * Get the last error message.
   *
   * @return the error message, or null if no error
   */
  public String getErrorMessage() {
    return errorMessage;
  }

  /**
   * Reset the parser to its initial state.
   */
  public void reset() {
    state = State.NEED_MORE_DATA;
    buffer.reset();
    haveMessageSize = false;
    messageSize = 0;
    bytesConsumed = 0;
    varintValue = 0;
    varintShift = 0;
    errorMessage = null;
    parsedMessage = null;
  }

  /**
   * Mark parsing as done (no more messages expected).
   */
  public void finish() {
    if (state == State.NEED_MORE_DATA && buffer.size() == 0) {
      state = State.DONE;
    }
  }
}
